### PR TITLE
New dim reduction

### PR DESCRIPTION
**Pull request by pablo-lauret**
## Scope
This PR modifies the dimensionality reduction module to add a functionality that allows the user to fit a dimensionality reduction object on a set of data X and then transform a set of data Y with this fitted object. Previously, this module only possessed a fit_transform functionality, that would fit on data X then transform data X, which is insufficient for our updated experiment pipeline for our alpha 3 and 4 models. 

Given that sklearn dim reduction objects work with numpy arrays, it made sense to change our embedder module to save embeddings as numpy arrays to a dataframe, and save the resulting vectorised dataset to a pickle file, which preserves python data types contrary to TSV, which we were using before. This PR makes the necessary modifications to the embedder module to ensure seamless interaction with the dimensionality reduction module.

## Implementation
New method implemented on dim_reduction.py, called 'fit'. This takes a dataframe\list of dataframes with vector embeddings and fits the dimensionality reduction object on the embeddings contained in those.

The method reduce_dimension has been modified to perform a transform operation, instead of fit_transform which it was performing before. This means that we can now fit on train+val, and transform train, val and test separately. 

The save_dataset methods in the embedder and dim_reduction modules have been modified to save dataframes as pickle files.

A number of methods in embedder and dim_reduction have received minor changes to ensure all embedding vectors are stored and read as numpy arrays.

## How to Test
Run the test_dim_reduction.py and test_embedder.py test modules. These have been modified to ensure the new changes are working properly.

## Emoji Guide
**For reviewers: Emojis can be added to comments to call out blocking versus non-blocking feedback.**
E.g: Praise, minor suggestions, or clarifying questions that don’t block merging the PR.
> 🟢 Nice refactor!
> 🟡 Why was the default value removed?
E.g: Blocking feedback must be addressed before merging.
> 🔴 This change will break something important

| | | |
| --- | --- | --- |
| Blocking | 🔴 ❌ 🚨 | Red |
| Non-blocking | 🟡 💡 🤔 💭 | Yellow, thinking, etc |
| Praise | 🟢 💚 😍 👍 🙌 | Green, hearts, positive emojis, etc |

### CODE REVIEW

**Comment by pablo-lauret** [comment]
**NOTE:** as agreed with the rest of the team, TSNE has been dropped from the dimensionality reduction module as it only  has fit and fit_transform methods, and no transform method, making it incompatible with our alpha3/4 pipeline's needs.

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 1]
```diff
@@ -4,275 +4,401 @@
 Module containing the base class for performing dimensionality reduction.
 
 """
-from abc import ABC, abstractmethod
 
-import numpy as np 
+import numpy as np
+import os
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py:8]: order of imports needs change. group them by style and within style alphab

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 2]
```diff
@@ -4,275 +4,401 @@
...
+        try:
+            self.embeddings_to_reduce = data_to_reduce[
+                "sentence_vectorised"
+            ].to_list()
+        except KeyError:
+            raise KeyError(
+                "Sentence vectors not present in the data to be reduced."
+            )
+        raise NotImplementedError(
+            "Suclasses must implement the reduce_dimension method."
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py:145]: typo

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 3]
```diff
@@ -4,275 +4,401 @@
...
-            'sentence_vector', with the vector representation 
-            of each sentence.
-        dim_out : int 
+        dim_out : int
             Desired output dimension of the vectors.
-        **kwargs 
+        **kwargs
             Arguments passed to the sklearn.decomposition.PCA object.
-            They can be found in 
+            They can be found in
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py]: arguments blbablabla can be found at xxx.
no peridod in the middle

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 4]
```diff
@@ -4,275 +4,401 @@
...
+            embeddings_to_fit = []
+            for df in data_to_fit:
+                df_embeddings = df["sentence_vectorised"].to_list()
+                embeddings_to_fit.extend(df_embeddings)
+            embeddings_to_fit = np.array(embeddings_to_fit)
+        else:
+            raise ValueError(
+                "Invalid input type. Expecting pd.DataFrame or list[pd.DataFrame]."
+            )
+        X = np.stack(embeddings_to_fit)
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py]: variable name: what's X?

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 5]
```diff
@@ -4,275 +4,401 @@
...
+        Reduces the embeddings in the input dataframe with the fitted
+        PCA.
         """
-        sentence_vectors_reduced = self.pca_sk.fit_transform(
-            self.sentence_vectors)
-        self.dataset[
-            'reduced_sentence_embedding'] = sentence_vectors_reduced.tolist()
+        reduced_data = data_to_reduce.copy()
+        embeddings_to_reduce = data_to_reduce["sentence_vectorised"].to_numpy()
+        X = np.stack(embeddings_to_reduce)
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py]: SAA

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 6]
```diff
@@ -4,275 +4,401 @@
...
+    def reduce_dimension(self, data_to_reduce: pd.DataFrame) -> pd.DataFrame:
+        """
+        Reduces the embeddings in the input dataframe with the fitted
+        PCA.
         """
-        sentence_vectors_reduced = self.pca_sk.fit_transform(
-            self.sentence_vectors)
-        self.dataset[
-            'reduced_sentence_embedding'] = sentence_vectors_reduced.tolist()
+        reduced_data = data_to_reduce.copy()
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py]: If we care: make sure you don't need a deepcopy.
If we don't care: don't copy just use in place

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 7]
```diff
@@ -4,275 +4,401 @@
...
+            embeddings_to_fit = []
+            for df in data_to_fit:
+                df_embeddings = df["sentence_vectorised"].to_list()
+                embeddings_to_fit.extend(df_embeddings)
+            embeddings_to_fit = np.array(embeddings_to_fit)
+        else:
+            raise ValueError(
+                "Invalid input type. Expecting pd.DataFrame or list[pd.DataFrame]."
+            )
+        X = np.stack(embeddings_to_fit)
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py]: SAA

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 8]
```diff
@@ -4,275 +4,401 @@
...
+        self.ica_sk.fit(X)
+
+    def reduce_dimension(self, data_to_reduce: pd.DataFrame) -> pd.DataFrame:
+        """
+        Reduces the embeddings in the input dataframe with the fitted
+        ICA.
+        """
+        reduced_data = data_to_reduce.copy()
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py]: SAA

File: neasqc_wp61/models/quantum/dim_reduction.py
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py:1]: Move the common parts of the code to the parent class.
Smth like 
pre_bla(...)
and
post_bla(...)

and call them from child class smth like:
def my_child_func():
super.pre_bla(x,y,z)
do child stuff
super.post_bla()

File: neasqc_wp61/models/quantum/embedder.py [hunk 9]
```diff
@@ -126,7 +135,7 @@ def _add_embeddings_to_dataset(
                 "compute_embeddings must be called before add_embeddings_to_dataset"
             )
         if not isinstance(embeddings, list):
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/embedder.py:140]: You can't have code + a raiseNotImplemented error.
Either you DO want your code to be run, in which case you need to call super, and you don't want raisNotImplemented to bet there! 
Or you you DON'T want your code to be run, and you raise a NIE.

Here in your code what's supposed to happen to if not isinstance???

File: neasqc_wp61/models/quantum/embedder.py [hunk 10]
```diff
@@ -143,7 +152,7 @@ def vectorise_dataset(self) -> None:
     @abstractmethod
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/embedder.py]: abstract???

File: neasqc_wp61/models/quantum/embedder.py [hunk 11]
```diff
@@ -153,8 +162,8 @@ def save_embedding_dataset(self, path: str, filename: str) -> None:
             include a / at the end).
         filename: str
             The name with which the generated dataset containing the
-            embeddings will be saved (it should not include .csv or any
-            other file extension).
+            embeddings will be saved (it should not include .pkl or
+            any other file extension).
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/embedder.py:164]: SAA you've got to choose between NIE and doing stuff

File: neasqc_wp61/models/quantum/embedder.py [hunk 12]
```diff
@@ -506,14 +522,19 @@ def _compute_embeddings(
...
         if self.is_sentence_embedding:
-            vectorised_sentence_list = model.encode(sentences).tolist()
+            vectorised_sentence_array = model.encode(
+                sentences, convert_to_numpy=True
+            )
+            vectorised_sentence_list = [
+                array.flatten() for array in vectorised_sentence_array
+            ]
 
         else:
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/embedder.py:442]: dangerous leap of faith here. 
This else means basically anything else, so it could be a potatoe for all you know.
Have we made sure earlier on that this was checked?

File: tests/test_embedder.py [hunk 13]
```diff
@@ -74,96 +74,74 @@ def check_dimension(
...
         True if all vectors have the correct dimension, False otherwise.
     """
-    for vector in vectors:
-        if is_sentence_embedding:
-            if len(vector) != dim:
-                return False
-        else:
-            for element in vector:
-                if len(element) != dim:
+    if is_sentence_embedding:
```
Comment (ryk-wolf) [inline tests/test_embedder.py:77]: the name of this isnt great. check that dimension... what? is correct? is small? is big? is present?

File: tests/test_embedder.py [hunk 14]
```diff
@@ -74,96 +74,74 @@ def check_dimension(
...
+                for element in vector:
+                    if element.shape[0] != dim:
+                        return False
+            else:
+                return False
     return True
 
 
-def check_sentence_embedding_type(x: Union[list, pd.Series]) -> bool:
+def check_sentence_embedding_type(x: list) -> bool:
```
Comment (ryk-wolf) [inline tests/test_embedder.py]: SAA check that type what..? is correct? is str? is smth else?

File: tests/test_embedder.py [hunk 15]
```diff
@@ -74,96 +74,74 @@ def check_dimension(
...
-        if isinstance(embedding, list):
-            return all(isinstance(element, float) for element in embedding)
+        if isinstance(embedding, np.ndarray):
+            return np.all(np.issubdtype(embedding.dtype, np.float32))
 
     return False
 
 
-def check_word_embedding_type(x: Union[list, pd.Series]) -> bool:
+def check_word_embedding_type(x: list) -> bool:
```
Comment (ryk-wolf) [inline tests/test_embedder.py]: SAA check that word embeddings do what???

**Review by ryk-wolf — CHANGES_REQUESTED** [review]

**Comment by pablo-lauret** [comment]
All comments adressed.

File: neasqc_wp61/models/quantum/dim_reduction.py (hunk 2 above)
Comment (ryk-wolf) [reply neasqc_wp61/models/quantum/dim_reduction.py:145]: typo still there but never mind xD

**Review by ryk-wolf — APPROVED** [review]
